Services/communication_service.c

```diff
@@ -37,155 +38,205 @@ SAT_returnState communication_service_app(csp_packet_t *packet) {
-      memcpy(&packet->data[OUT_DATA_BYTE], &S_config.S_paPower, sizeof(uint32_t));
+      memcpy(&packet->data[OUT_DATA_BYTE], &S_config.PA_Power, sizeof(uint32_t));
       set_packet_length(packet, sizeof(int8_t) + sizeof(uint32_t) + 1);
       if (queue_response(packet) != SATR_OK) {
         return SATR_ERROR;
       }
       break;
 
-    case GET_FirmwareV:
-        HAL_S_getFirmwareV(&S_Firmware.Version);
-        S_Firmware.Version = csp_hton32(S_Firmware.Version);// check if 32 is correct/necessary
-        status = 0;
-        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
-        memcpy(&packet->data[OUT_DATA_BYTE], &S_Firmware.Version, sizeof(float));
-        set_packet_length(packet, sizeof(int8_t) + sizeof(float) + 1);
-        if (queue_response(packet) != SATR_OK) {
-          return SATR_ERROR;
-        }
-        break;
-
     case GET_STATUS:
-      HAL_S_getStatus (&S_status.PWRGD, &S_status.TXL);
-      S_status.PWRGD = csp_hton32(S_status.PWRGD);
-      S_status.TXL = csp_hton32(S_status.TXL);
-      status = 0;
+      status = HAL_S_getStatus (&S_status);
+      if (sizeof(S_status) + 1 > csp_buffer_data_size()) {
+        return SATR_ERROR;
+      }
+      S_status.PWRGD = csp_hton32((uint32_t)S_status.PWRGD);
+      S_status.TXL = csp_hton32((uint32_t)S_status.TXL);
       memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
-      memcpy(&packet->data[OUT_DATA_BYTE], &S_status.PWRGD, sizeof(uint8_t));
-      memcpy(&packet->data[OUT_DATA_BYTE + sizeof(uint8_t)], &S_status.TXL, sizeof(uint8_t));
-      set_packet_length(packet, sizeof(int8_t) + 2*sizeof(uint8_t) + 1);
+      memcpy(&packet->data[OUT_DATA_BYTE], &S_status, sizeof(S_status));
+      set_packet_length(packet, sizeof(int8_t) + sizeof(S_status) + 1);
 
       if (queue_response(packet) != SATR_OK) {
         return SATR_ERROR;
       }
       break;
 
     case GET_TR:
-        HAL_S_getTR(&S_buffer.transmit);
-        S_buffer.transmit = csp_hton32(S_buffer.transmit);
+        status = HAL_S_getTR(&S_TR);
+        if (sizeof(S_TR) + 1 > csp_buffer_data_size()) {
+          return SATR_ERROR;
+        }
+        S_TR.transmit = csp_hton32((uint32_t)S_TR.transmit);
         // what to do for int?
-        status = 0;
         memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
-        memcpy(&packet->data[OUT_DATA_BYTE], &S_buffer.transmit, sizeof(int));
-        set_packet_length(packet, sizeof(int8_t) + sizeof(int) + 1);
+        memcpy(&packet->data[OUT_DATA_BYTE], &S_TR, sizeof(S_TR));
+        set_packet_length(packet, sizeof(int8_t) + sizeof(S_TR) + 1);
         if (queue_response(packet) != SATR_OK) {
           return SATR_ERROR;
         }
         break;
 
+    case GET_HK:
+        status = HAL_S_getHK (&S_HK);
+        if (sizeof(S_HK) + 1 > csp_buffer_data_size()) {
+          return SATR_ERROR;
+        }
+        /*Use va_list from stdarg for various number of arguments. */
```

Comment:
I tried but it became too complex, especially for the get_full_status because the types are different. If other services would want to send big structures or arrays like these ones, it'd be better to figure out how to define a generalized csp_hton function that calls the relative function based on the type of the arguments and accepts any number of them.

Reply:
Yeah, that would get complicated. I looked into this, I'm not sure if that's possible without doing some whacky function overloading. This is the best we can do for now.